### PR TITLE
Popover for remove copilot and change access level

### PR DESCRIPTION
**Pull request by rushatgabhane**
### Fixed Issues

$ https://github.com/Expensify/App/issues/46921
$ https://github.com/Expensify/App/issues/46925
PROPOSAL:

### Tests

Pre-req: be on copilot beta.

### Add copilot
1. Go to settings > security 
2. Add copilot
3. Set access level as limited

### Change access level
1. Go to settings > security
2. Click on an existing copilot
3. Verify you see a popover menu
4. Click change access level
5. Change access level to full
6. Enter magic code
7. Verify access level has changed.

### Remove copilot
1. Go to settings > security
2. Click on an existing copilot
3. Verify you see a popover menu
4. Click remove copilot
5. Verify you see a modal
6. Confirm
7. Verify that copilot is removed.

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/9e171fdb-1c39-41fc-8968-c90bd7c832db

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/c90c778f-9708-4779-a008-3aa55776637a

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/380260ab-96e5-4b21-aeaa-58d998a4cec5
</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

image

https://github.com/user-attachments/assets/f69b7942-412f-426b-9da5-954e8bb67b0b

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/bede10d8-65ee-42c0-8a1a-e196c2477d7e

</details>

### CODE REVIEW

**Comment by dangrous** [comment]
Posted on the issue too but for clarity - Backend is ready! UpdateDelegateRole now exists.

Still working on those other changes (error handling, and personal details for delegated access) but that shouldn't block this.

**Comment by rushatgabhane** [comment]
thank you! let me update and test this PR

**Comment by dangrous** [comment]
how are we looking on this one? Would love to get this settled in the next day or so!

**Comment by dangrous** [comment]
Bumping on this - thanks! Can we get this merged on Monday? It's the last big piece we have!

**Comment by rushatgabhane** [comment]
PR is ready for a review 
cc: @situchan

**Comment by situchan** [comment]
> Hey @situchan , I would be reviewing this PR because it belongs to the co-pilot project, thanks :)

Sure. Can you share link to that discussion (you were assigned to this project) though?

**Review by allgandalf — COMMENTED** [review]
Overall seems good to me, only a few comments, I will begin testing now

**Review by allgandalf — CHANGES_REQUESTED** [review]
Functionality wise this LGTM,  Lets work on the bugs and get this one merged ASAP 🚀

File: src/languages/es.ts [hunk 1]
```diff
@@ -5372,8 +5372,13 @@ const translations = {
                     return '';
             }
         },
+        removeCopilot: 'Eliminar copiloto',
```
[the remark this answers is not in the document]
Comment (rushatgabhane) [reply src/languages/es.ts:5375]: yes

File: src/libs/actions/Delegate.ts [hunk 2]
```diff
@@ -301,6 +301,72 @@ function addDelegate(email: string, role: DelegateRole, validateCode: string) {
     API.write(WRITE_COMMANDS.ADD_DELEGATE, parameters, {optimisticData, successData, failureData});
 }
 
+function removeDelegate(email: string) {
+    if (!email) {
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (rushatgabhane) [reply src/libs/actions/Delegate.ts:307]: makes it easier. we can use optional chaining

File: src/pages/settings/Security/AddDelegate/UpdateDelegateRole/ValidateCodeForm/index.android.tsx [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import React, {forwardRef} from 'react';
```
[the remark this answers is not in the document]
Comment (rushatgabhane) [reply src/pages/settings/Security/AddDelegate/UpdateDelegateRole/ValidateCodeForm/index.android.tsx:1]: yeah, android has different autocomplete prop

**Comment by rushatgabhane** [comment]
> Bug: The popover icons are missing the green color

@allgandalf i think it is correct. im reusing the component so the styles are same in other popovers too.

@Expensify/design could you please confirm if this design for popover color and icon fill is correct


https://github.com/user-attachments/assets/16dc2488-555d-4c52-8e53-02d3f7ce2cbe

**Comment by rushatgabhane** [comment]
@allgandalf all fixed

**Comment by dannymcclain** [comment]
>could you please confirm if this design for popover color and icon fill is correct

Yup! Looks good 👍

**Comment by dangrous** [comment]
Got some conflicts now, sounds like we're almost there though! Are we on track to final review/merge today or tomorrow?

**Comment by rushatgabhane** [comment]
@allgandalf all yours

**Comment by rushatgabhane** [comment]
hmmm, tried Modal.close but it didn't work

**Comment by dangrous** [comment]
How's the investigation going/

**Review by allgandalf — APPROVED** [review]
Lets merge this 🚀

**Comment by dangrous** [comment]
Awesome! Thanks for sussing that out

**Review by allgandalf — APPROVED** [review]
Tested on iOS native, works great

**Review by dangrous — APPROVED** [review]
Okay let's do it! I found a very edge case bug where the popover shows up in a weird place if you have changed your screen size on web since loading the page / opening it last, but we have a number of edge case bugs that exist about changing screen size so I don't think it's a blocker, not really a thing we specifically support.

Thank you both for all your hard work on this!

I'm going to merge, and let's just keep an eye out for any additional bugs

File: src/pages/settings/Security/AddDelegate/UpdateDelegateRole/UpdateDelegateRolePage.tsx [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                    </Text>
+                }
+                onSelectRow={(option) => {
+                    if (option.isSelected) {
+                        Navigation.dismissModal();
+                        return;
+                    }
+
+                    requestValidationCode();
+                    setCurrentRole(option.value);
```
Comment (alitoshmatov) [inline src/pages/settings/Security/AddDelegate/UpdateDelegateRole/UpdateDelegateRolePage.tsx:76]: When the option is selected user is required to authenticate using magic code, which means this option is not selected yet. Thus when user just goes back without authenticating this option will be selected even though user did not complete required steps. https://github.com/Expensify/App/pull/53677/files

File: src/libs/actions/Delegate.ts [hunk 5]
```diff
@@ -342,6 +408,144 @@ function removePendingDelegate(email: string) {
...
+            value: {
+                delegatedAccess: {
+                    delegates: delegatedAccess.delegates.map((delegate) =>
+                        delegate.email === email
+                            ? {
+                                  ...delegate,
+                                  role,
+                                  errorFields: {updateDelegateRole: null},
+                                  pendingAction: CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE,
+                                  pendingFields: {role: CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE},
```
Comment (FitseTLT) [inline src/libs/actions/Delegate.ts:429]: We should have cleared pending fields on `failureData` as the above code caused https://github.com/Expensify/App/issues/72381

**Pull request merged**